src/transformers/models/vits/modeling_vits.py

```diff
@@ -0,0 +1,1413 @@
+
+        # Initialize weights and apply final processing
+        self.post_init()
+
+    def get_encoder(self):
+        return self.text_encoder
+
+    @add_start_docstrings_to_model_forward(VITS_INPUTS_DOCSTRING)
+    @replace_return_docstrings(output_type=VitsModelOutput, config_class=_CONFIG_FOR_DOC)
+    def forward(
+        self,
+        input_ids: Optional[torch.Tensor] = None,
+        attention_mask: Optional[torch.Tensor] = None,
+        speed: float = 1.0,
+        noise_scale: float = 1.0,
+        noise_scale_duration: float = 1.0,
+        speaker_id: Optional[int] = None,
+        return_dict: Optional[bool] = None,
+        labels: Optional[torch.FloatTensor] = None,
+    ) -> Union[Tuple[torch.FloatTensor], VitsModelOutput]:
+        r"""
+        labels (`torch.FloatTensor` of shape `(batch_size, config.spectrogram_bins, sequence_length)`, *optional*):
+            Float values of target spectrogram. Timesteps set to `-100.0` are ignored (masked) for the loss
+            computation.
+
+        Returns:
+
+        Example:
+
+        ```python
+        >>> from transformers import VitsMmsTokenizer, VitsModel, set_seed
+        >>> import torch
+
+        >>> tokenizer = VitsMmsTokenizer.from_pretrained("Matthijs/mms-tts-eng")
+        >>> model = VitsModel.from_pretrained("Matthijs/mms-tts-eng")
+
+        >>> inputs = tokenizer(text="Hello, my dog is cute", return_tensors="pt")
+
+        >>> set_seed(555)  # make deterministic
+
+        >>> # generate speech
+        >>> outputs = model(inputs["input_ids"])
+        >>> outputs.audio.shape
+        torch.Size([1, 1, 33280])
+        ```
+        """
+        return_dict = return_dict if return_dict is not None else self.config.use_return_dict
+
+        if attention_mask is not None:
+            input_padding_mask = attention_mask.unsqueeze(-1).float()
+        else:
+            input_padding_mask = torch.ones_like(input_ids).unsqueeze(-1).float()
+
+        # Workaround for tokenizer: filter out padding tokens.
+        input_ids[input_ids >= self.config.vocab_size] = 0
+
+        hidden_states, means_prior, log_variances_prior = self.text_encoder(
+            input_ids, input_padding_mask, attention_mask
+        )
+
+        hidden_states = hidden_states.transpose(1, 2)
+        input_padding_mask = input_padding_mask.transpose(1, 2)
+
+        if self.config.num_speakers > 1:
+            if speaker_id is None:
+                raise ValueError("Expected speaker_id")
+            speaker_embeddings = self.embed_speaker(torch.tensor([speaker_id])).unsqueeze(-1)
+        else:
+            speaker_embeddings = None
+
+        if labels is not None:
+            raise NotImplementedError("Training of VITS is not supported yet.")
+
+        if self.config.use_stochastic_duration_prediction:
+            log_duration = self.duration_predictor(
+                hidden_states, input_padding_mask, speaker_embeddings, reverse=True, noise_scale=noise_scale_duration
```

Comment:
Is `reverse=False` only done when training? Maybe we can set `reverse=not self.training`? Probably fine though to leave it pinned to `reverse=True` and raise a `NotImplementedError` for now if labels are passed (thus blocking training)

Reply:
The training code performs different logic. I removed it because it requires the monotonic-align package, but there's a whole bunch of different code that goes in the `if labels is not None` block. Everything that's below that block is used only for inference. I think it's misleading to write `reverse=not self.training` since if will never be used for training.